network/validator/pubsub/authorized_sender_validator.go

```diff
@@ -0,0 +1,151 @@
+	authorizedRolesMap[cborcodec.CodeBlockVote] = flow.RoleList{flow.RoleConsensus}
+
+	// protocol state sync
+	authorizedRolesMap[cborcodec.CodeSyncRequest] = flow.Roles()
+	authorizedRolesMap[cborcodec.CodeSyncResponse] = flow.Roles()
+	authorizedRolesMap[cborcodec.CodeRangeRequest] = flow.Roles()
+	authorizedRolesMap[cborcodec.CodeBatchRequest] = flow.Roles()
+	authorizedRolesMap[cborcodec.CodeBlockResponse] = flow.Roles()
+
+	// cluster consensus
+	authorizedRolesMap[cborcodec.CodeClusterBlockProposal] = flow.RoleList{flow.RoleCollection}
+	authorizedRolesMap[cborcodec.CodeClusterBlockVote] = flow.RoleList{flow.RoleCollection}
+	authorizedRolesMap[cborcodec.CodeClusterBlockResponse] = flow.RoleList{flow.RoleCollection}
+
+	// collections, guarantees & transactions
+	authorizedRolesMap[cborcodec.CodeCollectionGuarantee] = flow.RoleList{flow.RoleCollection}
+	authorizedRolesMap[cborcodec.CodeTransactionBody] = flow.RoleList{flow.RoleCollection}
+	authorizedRolesMap[cborcodec.CodeTransaction] = flow.RoleList{flow.RoleCollection}
+
+	// core messages for execution & verification
+	authorizedRolesMap[cborcodec.CodeExecutionReceipt] = flow.RoleList{flow.RoleExecution}
+	authorizedRolesMap[cborcodec.CodeResultApproval] = flow.RoleList{flow.RoleVerification}
+
+	// execution state synchronization
+	// NOTE: these messages have been deprecated
+	authorizedRolesMap[cborcodec.CodeExecutionStateSyncRequest] = flow.RoleList{}
+	authorizedRolesMap[cborcodec.CodeExecutionStateDelta] = flow.RoleList{}
+
+	// data exchange for execution of blocks
+	authorizedRolesMap[cborcodec.CodeChunkDataRequest] = flow.RoleList{flow.RoleVerification}
+	authorizedRolesMap[cborcodec.CodeChunkDataResponse] = flow.RoleList{flow.RoleExecution}
+
+	// result approvals
+	authorizedRolesMap[cborcodec.CodeApprovalRequest] = flow.RoleList{flow.RoleConsensus}
+	authorizedRolesMap[cborcodec.CodeApprovalResponse] = flow.RoleList{flow.RoleVerification}
+
+	// generic entity exchange engines
+	authorizedRolesMap[cborcodec.CodeEntityRequest] = flow.RoleList{flow.RoleAccess, flow.RoleConsensus, flow.RoleCollection} // only staked access nodes
+	authorizedRolesMap[cborcodec.CodeEntityResponse] = flow.RoleList{flow.RoleCollection, flow.RoleExecution}
+
+	// testing
+	authorizedRolesMap[cborcodec.CodeEcho] = flow.Roles()
+
+	// dkg
+	authorizedRolesMap[cborcodec.CodeDKGMessage] = flow.RoleList{flow.RoleConsensus} // sn nodes for next epoch
+}
+
+// AuthorizedSenderValidator using the getIdentity func will check if the role of the sender
+// is part of the authorized roles list for the type of message being sent. A node is considered
+// to be authorized to send a message if all of the following are true.
+// 1. The message type is a known message type (initialized in the authorizedRolesMap).
+// 2. The authorized roles list for the message type contains the senders role.
+// 3. The node has a weight > 0 and is not ejected
```

Comment:
after discussing with Alex, we can skip the weight/ejected checks here since it would be hard to match the message with a specific view/block outside of the application layer. The weight check will need to go into the application layer (or at least outside of the topic validator)

Reply:
https://github.com/onflow/flow-go/pull/2317/commits/079042c5ef705456cae43752bd375f29afdec63c